include/pstl_offload/internal/usm_memory_replacement.h

```diff
@@ -53,34 +55,57 @@ __get_offload_device()
 class __offload_policy_holder_type
 {
     using __set_active_device_func_type = void (*)(sycl::device*);
+    using __get_offload_device_func_type = sycl::device (*)();
 
   public:
     // Since the global object of __offload_policy_holder_type is static but the constructor
     // of the class is inline, we need to avoid calling static functions inside of the constructor
     // and pass the pointer to exact function as an argument to guarantee that the correct __active_device
     // would be stored in each translation unit
-    __offload_policy_holder_type(sycl::device&& __offload_device,
+    __offload_policy_holder_type(__device_selector_type __device_selector,
```

Comment:
I guess we can simply make this constructor template (in the same manner as a `sycl::device` constructor):

```suggestion
    template <typename _DeviceSelector>
    __offload_policy_holder_type(const _DeviceSelector& __device_selector,
```

Reply:
Could you please clarify the reason a little? I mean, we know the exact type, we are going to use only this exact type, but not want to mention it because...

Reply:
As far as I understand from the spec, device selector is not guaranteed to be pointer to function, the only thing we know about it - it satisfies the requirements of `Callable` as described [here](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:device-selection:~:text=The%20interface%20for%20a%20device%20selector%20is%20any%20object%20that%20meets%20the%20C%2B%2B%20named%20requirement%20Callable%2C%20taking%20a%20parameter%20of%20type%20const%20device%20%26%20and%20returning%20a%20value%20that%20is%20implicitly%20convertible%20to%20int.). It covers also other invocable objects (lambdas, pointers to member functions/objects, etc.)

I would prefer to be consistent with SYCL classes (device, queue) as take/pass the selector argument in the same manner.

Reply:
> As far as I understand from the spec, device selector is not guaranteed to be pointer to function, the only thing we know about it - it satisfies the requirements of Callable as described [here](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:device-selection:~:text=The%20interface%20for%20a%20device%20selector%20is%20any%20object%20that%20meets%20the%20C%2B%2B%20named%20requirement%20Callable%2C%20taking%20a%20parameter%20of%20type%20const%20device%20%26%20and%20returning%20a%20value%20that%20is%20implicitly%20convertible%20to%20int.).

Thank you, missed that.